ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,437 @@
+Because of this we can assume that most Ruby engineers using the SDK won't be immediately familiar
+with writing async code. Because of this we have decided to make async execution opt-in, keeping
+everything synchronous by default.
+
+We have also chosen an eager execution pattern where any `async` block is started immediately and
+performed until blocked before proceeding with the execution. This is a bit easier to get a hold of
+and eliminates the risk of forgetting to actually start an async block. However this means that
+we'll need to find a solution for unhandled promise rejections, which we'll cover separately in a
+more detailed workflow proposal.
+
+We will introduce a helper method `async` to execute a block of code asynchronously:
+
+```ruby
+require 'sum_activity'
+
+class AsyncWorkflow < Temporal::Workflow
+  def execute
+    activity_promise = async { workflow.execute_activity(SumActivity, 1, 2) }
+    workflow_promise = async { workflow.execute_workflow(ChildWorkflow, 'parent') }
+
+    # Block execution until both promises are resolved
+    workflow.wait_for(activity_promise, workflow_promise)
+
+    activity_promise.result # => 3
+    workflow_promise.result # => 'Hello, parent!'
+  end
+end
+```
+
+A bit more involved example:
+
+```ruby
+class AnotherAsyncWorkflow < Temporal::Workflow
+  def execute
+    promise_1 = async { wait_and_sum(1, 2, delay: 10) }
+    promise_2 = async { wait_and_sum(3, 4, delay: 15) }
+
+    # The same exact method can be called in a blocking fashion
+    wait_and_sum(5, 6, delay: 5) # => 11
+
+    # Block execution until any promise is resolved
+    async.wait_for(async.any(promise_1, promise_2))
+
+    promise_1.result # => 3
+
+    # Because of the longest delay this would block until resolved
+    promise_2.result # => 7
+  end
+
+  private
+
+  def wait_and_sum(a, b, delay:)
+    workflow.sleep(delay)
+    workflow.execute_activity(SumActivity, a, b)
+  end
+end
+```
+
+
+## Activity heartbeats & cancellations
+
+This approach uses `Thread#raise` that will eject the execution out of the current context. In cases
+where this is not acceptable we will allow to either wrap a critical code in an `activity.shield`
+block (that will not raise until the block has executed) or opt-out of this default behaviour
+altogether and use `activity.cancelled?` flag to determine when an activity was cancelled.
+
+```ruby
+class MyActivity < Temporal::Activity
+  def execute
+    loop do
+      sleep 1
+      activity.heartbeat('ping')
+    end
+  rescue Temporal::Error::ActivityCancelled
+    raise 'this activity is cancelled'
+  end
+end
+```
+
+Here is what it will look like from a workflow perspective:
+
+```ruby
+class MyWorkflow < Temporal::Activity
+  def execute
+    # Async blocks double as cancellation scopes
+    promise = async { workflow.execute_activity(LongRunningActivity) }
+
+    workflow.sleep(10)
+
+    # Blocking call
+    promise.cancel
```

Comment:
You'd want to "race" between the activity and the timer to avoid unhandled rejection.
I'd replace this example to avoid suggesting this pattern.